test/net_http_test.rb

```diff
@@ -0,0 +1,389 @@
+        Semian["http_localhost_#{port + 1}"].reset if Semian["http_localhost_#{port + 1}"]
+        Semian.destroy('http_localhost_#{port}')
+        Semian.destroy('http_localhost_#{port + 1}')
+        with_server(port: port, reset_semian_state: false) do
+          with_toxic(upstream_port: port, toxic_port: port + 1) do |name|
+            open_circuit!(toxic_port: port + 1, toxic_name: name)
+            close_circuit!(toxic_port: port + 1)
+          end
+        end
+      end
+      with_server(port: PORT, reset_semian_state: false) do
+        open_circuit!
+        Net::HTTP.get(URI("http://127.0.0.1:#{PORT}/")) # different endpoint, should not raise errors
+      end
+    end
+  end
+
+  def test_persistent_state_after_server_restart
+    with_semian_options(DEFAULT_SEMIAN_OPTIONS) do
+      port = PORT + 100
+      with_server(port: port) do
+        with_toxic(upstream_port: port, toxic_port: port + 1) do |name|
+          open_circuit!(toxic_port: port + 1, toxic_name: name)
+        end
+      end
+      with_server(port: port, reset_semian_state: false) do
+        with_toxic(upstream_port: port, toxic_port: port + 1) do |_|
+          assert_raises Net::CircuitOpenError do
+            Net::HTTP.get(URI("http://localhost:#{port + 1}/200"))
+          end
+        end
+      end
+    end
+  end
+
+  private
+
+  def with_semian_options(options)
+    orig_semian_options = Semian::NetHTTP.raw_semian_options
+    Semian::NetHTTP.raw_semian_options = options
+    yield
+  ensure
+    Semian::NetHTTP.raw_semian_options = orig_semian_options
+  end
+
+  def with_custom_errors(errors)
+    orig_errors = Semian::NetHTTP.exceptions
+    errors.each do |error|
+      Semian::NetHTTP.exceptions << error
+    end
+    yield
+  ensure
+    Semian::NetHTTP.exceptions = orig_errors
+  end
+
+  def get_subclasses(klass)
+    ObjectSpace.each_object(klass.singleton_class).to_a - [klass]
+  end
+
+  def open_circuit!(toxic_port: TOXIC_PORT, toxic_name: "semian_test_net_http")
+    Net::HTTP.start("localhost", toxic_port) do |http|
+      http.read_timeout = 0.1
+      uri = URI("http://localhost:#{toxic_port}/200")
+      http.raw_semian_options[:error_threshold].times do
+        # Cause error error_threshold times so circuit opens
+        Toxiproxy[toxic_name].downstream(:latency, latency: 150).apply do
+          request = Net::HTTP::Get.new(uri)
+          assert_raises Net::ReadTimeout do
+            http.request(request)
+          end
+        end
+      end
+    end
+  end
+
+  def close_circuit!(toxic_port: TOXIC_PORT)
+    http = Net::HTTP.new("localhost", toxic_port)
+    Timecop.travel(http.raw_semian_options[:error_timeout])
+    # Cause successes success_threshold times so circuit closes
+    http.raw_semian_options[:success_threshold].times do
+      response = http.get("/200")
+      assert(200, response.code)
+    end
+  end
+
+  def with_server(port: PORT, reset_semian_state: true)
```

Comment:
How about making `with_server` check the `Host` header of the HTTP request to make sure you aren't receiving connections for hosts you don't want to? That also means you can test we trigger circuits etc. in that case.